src/crypto/elliptic_common.cpp

```diff
@@ -168,13 +168,49 @@ namespace fc { namespace ecc {
         return (fp[0] << 24) | (fp[1] << 16) | (fp[2] << 8) | fp[3];
     }
 
-    bool public_key::is_canonical( const compact_signature& c ) {
-        return !(c.data[1] & 0x80)
-               && !(c.data[1] == 0 && !(c.data[2] & 0x80))
-               && !(c.data[33] & 0x80)
-               && !(c.data[33] == 0 && !(c.data[34] & 0x80));
+    bool is_fc_canonical( const compact_signature& c )
+    {
+      return !(c.data[1] & 0x80)
+          && !(c.data[1] == 0 && !(c.data[2] & 0x80))
+          && !(c.data[33] & 0x80)
+          && !(c.data[33] == 0 && !(c.data[34] & 0x80));
+    }
+
+    bool is_bip_0062_canonical( const compact_signature& c )
+    {
+       // N/2 is 0x7FFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF 5D576E73 57A4501D DFE92F46 681B20A0
+       // According to BIP-0062, s must be <= N/2
+       const static uint8_t n_2[32] = {
+          0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+          0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+          0x5D, 0x57, 0x6E, 0x73, 0x57, 0xA4, 0x50, 0x1D,
+          0xDF, 0xE9, 0x2F, 0x46, 0x68, 0x1B, 0x20, 0xA0
+       };
+
+       for( size_t i = 0; i < 31; ++i )
+       {
+          if( c.data[33 + i ] != n_2[i] )
```

Comment:
I would really strongly recommend that we use a bignum library rather than perform bespoke bitwise arithmetic. Does steemd not have access to a bignum library (libgmp, etc.) that could do this heavy lifting? Alternatively, the `secp256k1_fe` type may be suitable. The library definitely has a constant-time comparison (`secp256k1_fe_equal`); but I don't know whether it implements `>`.

Also, i'm not sure if it matters, but this is not a constant-time comparison. to do that you would want something like the following:

```c++
uint8_t greater = 0;
uint8_t less = 0;
uint8_t a, b;

for (size_t i = 1; i < 31; i++) {
  a = c.data[33+i];
  b = n_2[i];
  greater |= (b - a) & ~less;
  less |= (a - b) & ~greater;
}
return ((greater | ~less) >> 7) != 0 && c.data[64] <= n_2[31];
```

If you want to completely exclude timing leaks you'd need to rearrange the final compound conditional into something bitwise.

Reply:
> Also, i'm not sure if it matters, but this is not a constant-time comparison.

While you can make the argument that my algorithm is O(n) it is bounded about by n=31, same as yours. Mine can simply short circuit as soon as we see the first byte is < `0x7F`.

Any bignum library we use will require converting the signature to native endian for comparison. This algorithm skips all of that by doing a byte-wise comparison.

The naive, but correct solution would be this:

``` c++
const static boost::uint256_t n_2 = 0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0_cppui256;

boost::uint256_t sig = 0;

for( size_t = 0; i < 4; i++ )
{
   sig << 8;
   sig += boost::endian::little_to_native( *( uint64_t* )( c.data + 33 + ( i * 8 ) ) );
}

return sig <= n_2;
```

My preference would be readability. Even with thousands of extra signatures, the unit tests using the old canonical code only takes 2 more seconds to execute. As a percentage of our execution time, signature verification is not that long. I was previously told by Dan that it was significant overhead, but reindexing the blockchain with full signature verification maybe only added a minute of reindex time. At a 1 hour reindex time, the difference is almost not perceptible. We don't normally do signature verification during reindex. The big performance hit is while live and during a sync. Even then. undo states are the killer there and cause the order of magnitude slowdown.

Reply:
Sorry, as far as regards the comparison: the goal for me is not _performance_, it's _security_. Any comparison is always going to be O(n) in the length of the longer string. I was saying that your proposed implementation leaks timing information by short-circuiting. Not sure if that's even relevant here; just wanted to point it out.

> Any bignum library we use will require converting the signature to native endian for comparison

Yes. For a minimal performance penalty you would gain a lot in readability and safety.

Reply:
@goldibex Updates have been pushed to use boost uint265_t